diff --git a/src/drivers/SConscript b/src/drivers/SConscript
index eedb59f..8f5cd5d 100644
--- a/src/drivers/SConscript
+++ b/src/drivers/SConscript
@@ -8,24 +8,24 @@ src_local = []
 objs = []
 
 objs += SConscript("uart/pl011/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'pl011'))
+                   duplicate=0, variant_dir = join(bdir, 'pl011'))
 
 objs += SConscript("timer/sp804/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'timer'))
+                   duplicate=0, variant_dir = join(bdir, 'timer'))
 
 objs += SConscript("irq/pl190/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'vic'))
+                   duplicate=0, variant_dir = join(bdir, 'vic'))
 
 objs += SConscript("irq/gic/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'gic'))
+                   duplicate=0, variant_dir = join(bdir, 'gic'))
 
 objs += SConscript("irq/omap3/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'omap/intc'))
+                   duplicate=0, variant_dir = join(bdir, 'omap/intc'))
 
 objs += SConscript("uart/omap/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'omap/uart'))
+                   duplicate=0, variant_dir = join(bdir, 'omap/uart'))
 
 objs += SConscript("timer/omap/SConscript", exports = { 'env' : env },
-                   duplicate=0, build_dir = join(bdir, 'omap/timer'))
+                   duplicate=0, variant_dir = join(bdir, 'omap/timer'))
 
 Return('objs')
diff --git a/conts/baremetal/empty/SConstruct b/conts/baremetal/empty/SConstruct
index b70d69a..4889d8e 100644
--- a/conts/baremetal/empty/SConstruct
+++ b/conts/baremetal/empty/SConstruct
@@ -48,7 +48,7 @@ env = Environment(CC = config.toolchain_userspace + 'gcc',
 		          CPPFLAGS = '-include l4/config.h -include l4/macros.h -include l4/types.h')
 
 objs = SConscript('SConscript', exports = { 'env' : env },
-                  duplicate=0, build_dir = builddir)
+                  duplicate=0, variant_dir = builddir)
 
 Depends(objs, join(PROJROOT, CONFIG_H))
 prog = env.Program(join(builddir, 'main.elf'), objs)
diff --git a/SConstruct b/SConstruct
index 2abc190..58c983d 100644
--- a/SConstruct
+++ b/SConstruct
@@ -71,35 +71,35 @@ env = Environment(CC = config.toolchain_kernel + 'gcc',
 objects = []
 objects += SConscript('src/generic/SConscript',
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, 'generic'))
+                      variant_dir = join(builddir, 'generic'))
 
 objects += SConscript(join(join('src/glue', arch), 'SConscript'),
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, join('glue',arch)))
+                      variant_dir = join(builddir, join('glue',arch)))
 
 objects += SConscript(join(join('src/arch', arch), 'SConscript'),
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, join('arch', arch)))
+                      variant_dir = join(builddir, join('arch', arch)))
 
 objects += SConscript(join(join('src/arch', arch), join(subarch, 'SConscript')),
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, join(join('arch',arch), subarch)))
+                      variant_dir = join(builddir, join(join('arch',arch), subarch)))
 
 objects += SConscript('src/lib/SConscript',
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, 'lib'))
+                      variant_dir = join(builddir, 'lib'))
 
 objects += SConscript('src/api/SConscript',
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, 'api'))
+                      variant_dir = join(builddir, 'api'))
 
 objects += SConscript('src/drivers/SConscript',
                       exports = { 'env' : env, 'bdir' : 'driver/'}, duplicate = 0,
-                      build_dir = join(builddir, 'driver'))
+                      variant_dir = join(builddir, 'driver'))
 
 objects += SConscript(join(join('src/platform', platform), 'SConscript'),
                       exports = { 'env' : env }, duplicate = 0,
-                      build_dir = join(builddir, join('platform', platform)))
+                      variant_dir = join(builddir, join('platform', platform)))
 
 
 # Add builders for generating kernel linker scripts
