--- src/netlib.h	(revision 644)
+++ src/netlib.h	(working copy)
@@ -536,7 +536,11 @@
 extern  void    dump_request();
 extern  void    dump_addrinfo(FILE *dumploc, struct addrinfo *info,
 			      char *host, char *port, int family);
+#ifdef GENODE_BUILD
+static inline  void    start_timer(int time) { }
+#else
 extern  void    start_timer(int time);
+#endif
 extern  void    stop_timer();
 extern  void    cpu_start(int measure_cpu);
 extern  void    cpu_stop(int measure_cpu, float *elapsed);
