diff -urN ff007e.orig/src/ffconf.h ff007e/src/ffconf.h
--- ff007e.orig/src/ffconf.h	2009-11-03 00:44:22.000000000 +0100
+++ ff007e/src/ffconf.h	2012-07-09 13:08:03.000000000 +0200
@@ -40,7 +40,7 @@
 /* To enable string functions, set _USE_STRFUNC to 1 or 2. */
 
 
-#define	_USE_MKFS	0		/* 0 or 1 */
+#define	_USE_MKFS	1		/* 0 or 1 */
 /* To enable f_mkfs function, set _USE_MKFS to 1 and set _FS_READONLY to 0 */
 
 
@@ -53,7 +53,7 @@
 / Locale and Namespace Configurations
 /----------------------------------------------------------------------------*/
 
-#define _CODE_PAGE	932
+#define _CODE_PAGE	858
 /* The _CODE_PAGE specifies the OEM code page to be used on the target system.
 /  Incorrect setting of the code page can cause a file open failure.
 /
@@ -86,7 +86,7 @@
 */
 
 
-#define	_USE_LFN	0		/* 0, 1 or 2 */
+#define	_USE_LFN	1		/* 0, 1 or 2 */
 #define	_MAX_LFN	255		/* Maximum LFN length to handle (12 to 255) */
 /* The _USE_LFN option switches the LFN support.
 /
@@ -105,7 +105,7 @@
 */
 
 
-#define _FS_RPATH	0		/* 0 or 1 */
+#define _FS_RPATH	1		/* 0 or 1 */
 /* When _FS_RPATH is set to 1, relative path feature is enabled and f_chdir,
 /  f_chdrive function are available.
 /  Note that output of the f_readdir fnction is affected by this option. */
