diff --git a/builtins/Makefile.in b/builtins/Makefile.in
--- a/builtins/Makefile.in
+++ b/builtins/Makefile.in
@@ -68,7 +68,7 @@
 LDFLAGS_FOR_BUILD = @LDFLAGS_FOR_BUILD@ $(LOCAL_LDFLAGS) $(CFLAGS_FOR_BUILD)
 LOCAL_LDFLAGS = @LOCAL_LDFLAGS@
 #LIBS_FOR_BUILD = @LIBS_FOR_BUILD@
-LIBS_FOR_BUILD = $(LIBS)
+#LIBS_FOR_BUILD = $(LIBS)
 
 BASHINCDIR = ${topdir}/include
 
diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -28398,8 +28398,8 @@
 else
   if test "$cross_compiling" = yes; then
   { $as_echo "$as_me:$LINENO: WARNING: cannot check whether getcwd allocates memory when cross-compiling -- defaulting to no" >&5
-$as_echo "$as_me: WARNING: cannot check whether getcwd allocates memory when cross-compiling -- defaulting to no" >&2;}
-    bash_cv_getcwd_malloc=no
+$as_echo "$as_me: WARNING: cannot check whether getcwd allocates memory when cross-compiling -- defaulting to yes" >&2;}
+    bash_cv_getcwd_malloc=yes
 
 else
   cat >conftest.$ac_ext <<_ACEOF
diff --git a/lib/intl/l10nflist.c b/lib/intl/l10nflist.c
--- a/lib/intl/l10nflist.c
+++ b/lib/intl/l10nflist.c
@@ -33,7 +33,7 @@
 #include <string.h>
 
 #if defined _LIBC || defined HAVE_ARGZ_H
-# include <argz.h>
+//# include <argz.h>
 #endif
 #include <ctype.h>
 #include <sys/types.h>
diff --git a/shell.c b/shell.c
--- a/shell.c
+++ b/shell.c
@@ -364,6 +364,12 @@
   env = environ;
 #endif /* __OPENNT */
 
+  /*
+   * Noux: Overcome missing declaration of 'env' that gets tiggered by the
+   *       'NO_MAIN_ENV_ARG' definition.
+   */
+  char **env = environ;
+
   USE_VAR(argc);
   USE_VAR(argv);
   USE_VAR(env);
