diff -ru bash-4.1/builtins/Makefile.in _bash-4.1/builtins/Makefile.in
--- contrib/bash-4.1/builtins/Makefile.in	2009-05-30 06:34:30.000000000 +0200
+++ contrib/bash-4.1/builtins/Makefile.in	2011-02-02 22:03:49.000000000 +0100
@@ -68,7 +68,7 @@
 LDFLAGS_FOR_BUILD = @LDFLAGS_FOR_BUILD@ $(LOCAL_LDFLAGS) $(CFLAGS_FOR_BUILD)
 LOCAL_LDFLAGS = @LOCAL_LDFLAGS@
 #LIBS_FOR_BUILD = @LIBS_FOR_BUILD@
-LIBS_FOR_BUILD = $(LIBS)
+#LIBS_FOR_BUILD = $(LIBS)
 
 BASHINCDIR = ${topdir}/include
 
diff -ru bash-4.1/lib/intl/l10nflist.c _bash-4.1/lib/intl/l10nflist.c
--- contrib/bash-4.1/lib/intl/l10nflist.c	2009-01-04 20:32:30.000000000 +0100
+++ contrib/bash-4.1/lib/intl/l10nflist.c	2011-02-02 21:55:34.000000000 +0100
@@ -33,7 +33,7 @@
 #include <string.h>
 
 #if defined _LIBC || defined HAVE_ARGZ_H
-# include <argz.h>
+//# include <argz.h>
 #endif
 #include <ctype.h>
 #include <sys/types.h>
diff -u bash-4.1/shell.c /home/no/src/genode/ports/contrib/bash-4.1/shell.c
--- contrib/bash-4.1/shell.c	2009-11-19 16:05:54.000000000 +0100
+++ contrib/bash-4.1/shell.c	2011-02-23 14:09:24.000000000 +0100
@@ -364,6 +364,12 @@
   env = environ;
 #endif /* __OPENNT */
 
+  /*
+   * Noux: Overcome missing declaration of 'env' that gets tiggered by the
+   *       'NO_MAIN_ENV_ARG' definition.
+   */
+  char **env = environ;
+
   USE_VAR(argc);
   USE_VAR(argv);
   USE_VAR(env);
