--- src/app/lighttpd/src/fdevent.c
+++ src/app/lighttpd/src/fdevent.c
@@ -51,7 +51,7 @@
 }
 
 void fdevent_setfd_cloexec(int fd) {
-#ifdef FD_CLOEXEC
+#if 0
 	if (fd < 0) return;
 	force_assert(-1 != fcntl(fd, F_SETFD, FD_CLOEXEC));
 #else
@@ -60,7 +60,7 @@
 }
 
 void fdevent_clrfd_cloexec(int fd) {
-#ifdef FD_CLOEXEC
+#if 0
 	if (fd >= 0) force_assert(-1 != fcntl(fd, F_SETFD, 0));
 #else
 	UNUSED(fd);
@@ -96,7 +96,7 @@
 		return socket(domain, type | SOCK_CLOEXEC, protocol);
 #endif
 	if (-1 != (fd = socket(domain, type, protocol))) {
-#ifdef FD_CLOEXEC
+#if 0
 		force_assert(-1 != fcntl(fd, F_SETFD, FD_CLOEXEC));
 #endif
 	}
@@ -120,7 +120,7 @@
        #endif
 #endif
 	if (-1 != (fd = socket(domain, type, protocol))) {
-#ifdef FD_CLOEXEC
+#if 0
 		force_assert(-1 != fcntl(fd, F_SETFD, FD_CLOEXEC));
 #endif
 #ifdef O_NONBLOCK
@@ -173,7 +173,7 @@
 #endif /* not used */
 
 int fdevent_dup_cloexec (int fd) {
-  #ifdef F_DUPFD_CLOEXEC
+  #if 0
     return fcntl(fd, F_DUPFD_CLOEXEC, 3);
   #else
     const int newfd = fcntl(fd, F_DUPFD, 3);
@@ -203,7 +203,7 @@
 /* O_CLOEXEC handled further below, if defined) */
 #ifdef O_NONBLOCK
 #define FDEVENT_O_FLAGS \
-        (O_BINARY | O_LARGEFILE | O_NOCTTY | O_NONBLOCK)
+        (O_BINARY | O_LARGEFILE | O_NOCTTY )
 #else
 #define FDEVENT_O_FLAGS \
         (O_BINARY | O_LARGEFILE | O_NOCTTY )
@@ -214,7 +214,7 @@
 	return open(pathname, flags | O_CLOEXEC | FDEVENT_O_FLAGS, mode);
 #else
 	int fd = open(pathname, flags | FDEVENT_O_FLAGS, mode);
-#ifdef FD_CLOEXEC
+#if 0
 	if (fd != -1)
 		force_assert(-1 != fcntl(fd, F_SETFD, FD_CLOEXEC));
 #endif
@@ -254,7 +254,7 @@
   #endif
     {
         if (0 != pipe(fds)
-         #ifdef FD_CLOEXEC
+         #if 0
          || 0 != fcntl(fds[0], F_SETFD, FD_CLOEXEC)
          || 0 != fcntl(fds[1], F_SETFD, FD_CLOEXEC)
          #endif
@@ -378,7 +378,7 @@
 char ** fdevent_environ (void) { return environ; }
 
 
-#ifdef FD_CLOEXEC
+#if 0
 static int fdevent_dup2_close_clrfd_cloexec(int oldfd, int newfd) {
     if (oldfd >= 0) {
         if (oldfd != newfd) {
@@ -406,7 +406,7 @@
 
 
 int fdevent_set_stdin_stdout_stderr(int fdin, int fdout, int fderr) {
-  #ifdef FD_CLOEXEC
+  #if 0
     if (STDIN_FILENO != fdevent_dup2_close_clrfd_cloexec(fdin, STDIN_FILENO))
         return -1;
     if (STDOUT_FILENO != fdevent_dup2_close_clrfd_cloexec(fdout, STDOUT_FILENO))
