diff --git a/src/sys/arch/arm/include/Makefile.common b/src/sys/arch/arm/include/Makefile.common
index 6d7643a..c91f55c 100644
--- a/src/sys/arch/arm/include/Makefile.common
+++ b/src/sys/arch/arm/include/Makefile.common
@@ -2,7 +2,7 @@
 
 .PATH: ../../arm/include/common
 
-.if 0
+.if 1
 INCS+=	ansi.h aout_machdep.h asm.h \
 	bswap.h \
 	cdefs.h cpu.h \
diff --git a/src/sys/arch/evbarm/Makefile b/src/sys/arch/evbarm/Makefile
index c4340cb..ec9445b 100644
--- a/src/sys/arch/evbarm/Makefile
+++ b/src/sys/arch/evbarm/Makefile
@@ -3,7 +3,7 @@
 # Makefile for evbarm tags file and boot blocks
 
 # Find where ARM source files are for inclusion in tags
-.include <../arm/Makefile.inc>
+#.include <../arm/Makefile.inc>
 
 TEVBARM=	${SYSDIR}/arch/evbarm/tags
 SEVBARM=	${SYSDIR}/arch/evbarm/adi_brh/*.[ch]
