diff --git a/src/tabwidget.cpp b/src/tabwidget.cpp
--- a/src/tabwidget.cpp
+++ b/src/tabwidget.cpp
@@ -829,7 +829,7 @@
 
     if (url.scheme() == QLatin1String("about")
         && url.path() == QLatin1String("home"))
-        url = QUrl(QLatin1String("qrc:/startpage.html"));
+        url = QUrl(QLatin1String("qrc:/demo/intro.html"));
 
     // QUrl::isValid() is too much tolerant.
     // We actually want to check if the url conforms to the RFC, which QUrl::isValid() doesn't state.
diff --git a/src/webpage.cpp b/src/webpage.cpp
--- a/src/webpage.cpp
+++ b/src/webpage.cpp
@@ -211,7 +211,7 @@
         frame = qobject_cast<QWebFrame*>(sender());
 
         if (frame->url().scheme() == QLatin1String("qrc")
-            && frame->url().path() == QLatin1String("/startpage.html")) {
+            && frame->url().path() == QLatin1String("/demo/intro.html")) {
 
             if (!m_javaScriptAroraObject)
                 m_javaScriptAroraObject = new JavaScriptAroraObject(this);
