--- contrib/fuse-ext2-svn-220/fuse-ext2/fuse-ext2.c.orig	2013-11-27 17:49:58.000000000 +0100
+++ contrib/fuse-ext2-svn-220/fuse-ext2/fuse-ext2.c	2013-11-27 17:50:45.593877786 +0100
@@ -278,7 +278,7 @@
 	goto exit;
 }
 
-static const struct fuse_operations ext2fs_ops = {
+const struct fuse_operations ext2fs_ops = {
 	.getattr        = op_getattr,
 	.readlink       = op_readlink,
 	.mknod          = op_mknod,
@@ -317,6 +317,7 @@
 	.bmap           = NULL,
 };
 
+#if 0
 int main (int argc, char *argv[])
 {
 	int err = 0;
@@ -385,3 +386,4 @@
 	free(opts.volname);
 	return err;
 }
+#endif
