+++ src/app/virtualbox/src/VBox/VMM/VMMR3/TM.cpp
@@ -1901,7 +1901,7 @@
 #ifdef VBOX_WITH_REM
         REMR3NotifyTimerPending(pVM, pVCpuDst);
 #endif
-        VMR3NotifyCpuFFU(pVCpuDst->pUVCpu, VMNOTIFYFF_FLAGS_DONE_REM /** @todo | VMNOTIFYFF_FLAGS_POKE ?*/);
+        VMR3NotifyCpuFFU(pVCpuDst->pUVCpu, VMNOTIFYFF_FLAGS_DONE_REM | VMNOTIFYFF_FLAGS_POKE);
         STAM_COUNTER_INC(&pVM->tm.s.StatTimerCallbackSetFF);
     }
 }
