Fix GCC 8 compile error

diff --git a/user/include/l4/kip.h b/user/include/l4/kip.h
index 63e5b62..4a8bd6e 100644
--- a/user/include/l4/kip.h
+++ b/user/include/l4/kip.h
@@ -558,7 +558,7 @@ L4_INLINE L4_Bool_t L4_HasFeature (const char *feature_name)
     void *kip = L4_GetKernelInterface();
     char *name;
 
-    for( L4_Word_t i = 0; (name = L4_Feature(kip,i)) != '\0'; i++ )
+    for( L4_Word_t i = 0; (name = L4_Feature(kip,i)) != 0; i++ )
     {
         const char *n = name;
         const char *fn = feature_name;
