diff -u dash-0.5.6/config.h.in _dash-0.5.6/config.h.in
--- contrib/dash-0.5.6/config.h.in	2010-04-02 08:10:53.000000000 +0200
+++ contrib/dash-0.5.6/config.h.in	2011-02-02 23:02:10.000000000 +0100
@@ -116,6 +116,8 @@
 # undef _GNU_SOURCE
 #endif
 
+#ifdef HAVE_CONFIG_H
+
 /* 64-bit operations are the same as 32-bit */
 #undef fstat64
 
@@ -130,3 +132,5 @@
 
 /* 64-bit operations are the same as 32-bit */
 #undef stat64
+
+#endif /* HAVE_CONFIG_H */
