+++ arch/l4/kernel/arch-arm/vmlinux.lds.S
@@ -108,6 +108,12 @@ SECTIONS
 		LONG(0xffffffff);
 		_parent_cap_local_name = .;
 		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
 	} : rw
 
 	. = ALIGN(4096);
+++ arch/l4/kernel/arch-x86/vmlinux.lds.S
@@ -101,6 +101,13 @@ SECTIONS
 		LONG(0xffffffff);
 		_parent_cap_local_name = .;
 		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
 
 		/*
 		 * Platform-specific entry for Fiasco.OC.
