bfd.patch

From: Christian Prochaska <christian.prochaska@genode-labs.com>


---
 bfd/bfd-in2.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/bfd/bfd-in2.h b/bfd/bfd-in2.h
index 1b82b6bc..900221ec 100644
--- a/bfd/bfd-in2.h
+++ b/bfd/bfd-in2.h
@@ -46,6 +46,7 @@ extern "C" {
 #include <time.h>
 #include "diagnostics.h"
 #include <stdarg.h>
+#include <stdio.h>
 #include <string.h>
 #include <sys/stat.h>
 
