commit 3085e764d1545cd2dae199037265c8d5c8a1b458
Author: Sebastian Sumpf <sebastian.sumpf@genode-labs.com>
Date:   Mon Jun 2 13:25:18 2014 +0200

    icmp.patch

diff --git a/net/ipv4/icmp.c b/net/ipv4/icmp.c
index 0134663..aee0c67 100644
--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -822,7 +822,7 @@ static void icmp_echo(struct sk_buff *skb)
 	if (!net->ipv4.sysctl_icmp_echo_ignore_all) {
 		struct icmp_bxm icmp_param;
 
-		icmp_param.data.icmph	   = *icmp_hdr(skb);
+		memcpy(&icmp_param.data.icmph, icmp_hdr(skb), sizeof(struct icmphdr)); 
 		icmp_param.data.icmph.type = ICMP_ECHOREPLY;
 		icmp_param.skb		   = skb;
 		icmp_param.offset	   = 0;
