diff --git a/drivers/gpu/imx/dcss/dcss-blkctl.c b/drivers/gpu/imx/dcss/dcss-blkctl.c
index 2f13b33..a4b0620 100644
--- a/drivers/gpu/imx/dcss/dcss-blkctl.c
+++ b/drivers/gpu/imx/dcss/dcss-blkctl.c
@@ -81,7 +81,7 @@ void dcss_blkctl_cfg(struct dcss_soc *dcss)
 	struct dcss_blkctl_priv *blkctl = dcss->blkctl_priv;
 
 	if (blkctl->hdmi_output)
-		dcss_writel((blkctl->clk_setting ^ HDMI_MIPI_CLK_SEL),
+		dcss_writel((blkctl->clk_setting),
 		    blkctl->base_reg + DCSS_BLKCTL_CONTROL0);
 	else
 		dcss_writel((blkctl->clk_setting ^ HDMI_MIPI_CLK_SEL) |
