--- a/drivers/gpu/drm/i915/intel_panel.c
+++ a/drivers/gpu/drm/i915/intel_panel.c
@@ -1715,6 +1715,7 @@
 
 	panel->backlight.present = true;
 
+	intel_panel_set_backlight(connector->state, panel->backlight.max*75/100, panel->backlight.max);
 	DRM_DEBUG_KMS("Connector %s backlight initialized, %s, brightness %u/%u\n",
 		      connector->name,
 		      panel->backlight.enabled ? "enabled" : "disabled",
