diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 32cf973..ec1d558 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -15816,3 +15816,11 @@ void intel_modeset_preclose(struct drm_device *dev, struct drm_file *file)
 		spin_unlock_irq(&dev->event_lock);
 	}
 }
+
+struct drm_framebuffer *
+lx_c_intel_framebuffer_create(struct drm_device *dev,
+							   struct drm_mode_fb_cmd2 *mode_cmd,
+							   struct drm_i915_gem_object *obj) {
+	return intel_framebuffer_create(dev, mode_cmd, obj);
+}
+EXPORT_SYMBOL(lx_c_intel_framebuffer_create);
