size_t is unsigned long on Genode

--- src/linux/include/uapi/asm-generic/posix_types.h
+++ src/linux/include/uapi/asm-generic/posix_types.h
@@ -65,7 +65,7 @@ typedef unsigned int	__kernel_old_dev_t;
  */
 #ifndef __kernel_size_t
 #if __BITS_PER_LONG != 64
-typedef unsigned int	__kernel_size_t;
+typedef unsigned long	__kernel_size_t;
 typedef int		__kernel_ssize_t;
 typedef int		__kernel_ptrdiff_t;
 #else
