diff --git a/net/ipv4/af_inet.c b/net/ipv4/af_inet.c
index 5c5db66..6cb68a1 100644
--- a/net/ipv4/af_inet.c
+++ b/net/ipv4/af_inet.c
@@ -733,6 +733,8 @@
 	    inet_autobind(sk))
 		return -EAGAIN;
 
+	if (sk->sk_wq == 0) sk->sk_wq = sock->wq;
+
 	return sk->sk_prot->sendmsg(sk, msg, size);
 }
 EXPORT_SYMBOL(inet_sendmsg);
@@ -764,6 +766,8 @@
 
 	sock_rps_record_flow(sk);
 
+	if (sk->sk_wq == 0) sk->sk_wq = sock->wq;
+
 	err = sk->sk_prot->recvmsg(sk, msg, size, flags & MSG_DONTWAIT,
 				   flags & ~MSG_DONTWAIT, &addr_len);
 	if (err >= 0)
