diff --git a/drivers/input/input.c b/drivers/input/input.c
index 9785546..cf6f435 100644
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -1885,6 +1885,7 @@ void input_free_device(struct input_dev *dev)
 						devm_input_device_release,
 						devm_input_device_match,
 						dev));
+		mutex_destroy(&dev->mutex);
 		input_put_device(dev);
 	}
 }
@@ -2024,6 +2025,7 @@ static void __input_unregister_device(struct input_dev *dev)
 
 	mutex_unlock(&input_mutex);
 
+	mutex_destroy(&dev->mutex);
 	device_del(&dev->dev);
 }
 
