diff --git a/drivers/hid/usbhid/hid-core.c b/drivers/hid/usbhid/hid-core.c
index 77c50cd..55379da 100644
--- a/drivers/hid/usbhid/hid-core.c
+++ b/drivers/hid/usbhid/hid-core.c
@@ -1636,7 +1636,7 @@ struct usb_interface *usbhid_find_interface(int minor)
 	return usb_find_interface(&hid_driver, minor);
 }
 
-static int __init hid_init(void)
+static int __init usbhid_init(void)
 {
 	int retval = -ENOMEM;
 
@@ -1655,14 +1655,14 @@ usbhid_quirks_init_fail:
 	return retval;
 }
 
-static void __exit hid_exit(void)
+static void __exit usbhid_exit(void)
 {
 	usb_deregister(&hid_driver);
 	hid_quirks_exit(BUS_USB);
 }
 
-module_init(hid_init);
-module_exit(hid_exit);
+module_init(usbhid_init);
+module_exit(usbhid_exit);
 
 MODULE_AUTHOR("Andreas Gal");
 MODULE_AUTHOR("Vojtech Pavlik");
