--- contrib/e2fsprogs-v1.42.9/misc/Makefile.in.orig	2014-01-07 15:22:06.554085193 +0100
+++ contrib/e2fsprogs-v1.42.9/misc/Makefile.in	2014-01-07 14:13:32.026193047 +0100
@@ -84,7 +84,7 @@
 		$(srcdir)/ismounted.c $(srcdir)/../e2fsck/profile.c \
 		$(srcdir)/e2undo.c $(srcdir)/e2freefrag.c
 
-LIBS= $(LIBEXT2FS) $(LIBCOM_ERR) 
+LIBS= $(LIBEXT2FS) $(LIBCOM_ERR) @LIBS@
 DEPLIBS= $(LIBEXT2FS) $(DEPLIBCOM_ERR)
 PROFILED_LIBS= $(PROFILED_LIBEXT2FS) $(PROFILED_LIBCOM_ERR)
 PROFILED_DEPLIBS= $(PROFILED_LIBEXT2FS) $(DEPPROFILED_LIBCOM_ERR)
--- contrib/e2fsprogs-v1.42.9/ext2ed/Makefile.in.orig	2014-01-07 15:24:29.566081444 +0100
+++ contrib/e2fsprogs-v1.42.9/ext2ed/Makefile.in	2014-01-07 14:13:58.522192352 +0100
@@ -16,7 +16,7 @@
 
 DOC_DIR		=	$datadir/doc/ext2ed
 
-LIBS		=	-lncurses $(LIBEXT2FS)
+LIBS		=	-lncurses $(LIBEXT2FS) @LIBS@
 
 SRCS=	$(srcdir)/main.c $(srcdir)/init.c $(srcdir)/general_com.c       \
 	$(srcdir)/inode_com.c $(srcdir)/dir_com.c $(srcdir)/super_com.c \
--- contrib/e2fsprogs-v1.42.9/intl/Makefile.in.orig	2014-01-07 15:25:01.830080600 +0100
+++ contrib/e2fsprogs-v1.42.9/intl/Makefile.in	2014-01-07 12:29:48.842356174 +0100
@@ -191,7 +191,7 @@
 .c.o:
 	$(E) "	CC $<"
 	$(Q) $(COMPILE) $<
-	$(Q) $(CHECK_CMD) $(ALL_CFLAGS) $<
+	#$(Q) $(CHECK_CMD) $(ALL_CFLAGS) $<
 
 .y.c:
 	$(YACC) $(YFLAGS) --output $@ $<
--- contrib/e2fsprogs-v1.42.9/e2fsck/Makefile.in.orig	2014-01-07 15:26:48.000000000 +0100
+++ contrib/e2fsprogs-v1.42.9/e2fsck/Makefile.in	2014-01-07 14:14:31.250191497 +0100
@@ -16,7 +16,7 @@
 FMANPAGES=	e2fsck.conf.5
 
 LIBS= $(LIBQUOTA) $(LIBEXT2FS) $(LIBCOM_ERR) $(LIBBLKID) $(LIBUUID) \
-	$(LIBINTL) $(LIBE2P)
+	$(LIBINTL) $(LIBE2P) @LIBS@
 DEPLIBS= $(DEPLIBQUOTA) $(LIBEXT2FS) $(DEPLIBCOM_ERR) $(DEPLIBBLKID) \
 	 $(DEPLIBUUID) $(DEPLIBE2P)
 
--- contrib/e2fsprogs-v1.42.9/resize/Makefile.in.orig	2014-01-07 15:27:41.000000000 +0100
+++ contrib/e2fsprogs-v1.42.9/resize/Makefile.in	2014-01-07 14:14:39.522191277 +0100
@@ -28,7 +28,7 @@
 	$(srcdir)/resource_track.c \
 	$(srcdir)/sim_progress.c
 
-LIBS= $(LIBE2P) $(LIBEXT2FS) $(LIBCOM_ERR) $(LIBINTL)
+LIBS= $(LIBE2P) $(LIBEXT2FS) $(LIBCOM_ERR) $(LIBINTL) @LIBS@
 DEPLIBS= $(LIBE2P) $(LIBEXT2FS) $(DEPLIBCOM_ERR)
 
 STATIC_LIBS= $(STATIC_LIBE2P) $(STATIC_LIBEXT2FS) $(STATIC_LIBCOM_ERR) \
--- contrib/e2fsprogs-v1.42.9/debugfs/Makefile.in.orig	2014-01-07 15:28:14.000000000 +0100
+++ contrib/e2fsprogs-v1.42.9/debugfs/Makefile.in	2014-01-07 14:14:47.890191058 +0100
@@ -31,7 +31,7 @@
 	$(srcdir)/filefrag.c $(srcdir)/extent_inode.c $(srcdir)/zap.c
 
 LIBS= $(LIBEXT2FS) $(LIBE2P) $(LIBSS) $(LIBCOM_ERR) $(LIBBLKID) \
-	$(LIBUUID)
+	$(LIBUUID) @LIBS@
 DEPLIBS= $(LIBEXT2FS) $(LIBE2P) $(DEPLIBSS) $(DEPLIBCOM_ERR) \
 	$(DEPLIBBLKID) $(DEPLIBUUID)
 
