diff --git a/examples/webkit/previewer/previewer.pro b/examples/webkit/previewer/previewer.pro
--- a/examples/webkit/previewer/previewer.pro
+++ b/examples/webkit/previewer/previewer.pro
@@ -11,8 +11,3 @@
 sources.files = $$SOURCES $$HEADERS $$FORMS $$RESOURCES *.pro images
 sources.path = $$[QT_INSTALL_EXAMPLES]/webkit/previewer
 INSTALLS += target sources
-
-symbian {
-    TARGET.UID3 = 0xA000CF6F
-    include($$QT_SOURCE_TREE/examples/symbianpkgrules.pri)
-}
diff --git a/examples/webkit/previewer/previewer.pro b/examples/webkit/previewer/main.cpp
--- a/examples/webkit/previewer/main.cpp
+++ b/examples/webkit/previewer/main.cpp
@@ -47,6 +47,7 @@
 {
     QApplication app(argc, argv);
     MainWindow mainWindow;
+    mainWindow.move(200, 200);
     mainWindow.show();
     return app.exec();
 }
