- include time.h where time_t is used

diff --git a/src/3rdparty/webkit/WebCore/loader/icon/IconRecord.h b/src/3rdparty/webkit/WebCore/loader/icon/IconRecord.h
--- a/src/3rdparty/webkit/WebCore/loader/icon/IconRecord.h
+++ b/src/3rdparty/webkit/WebCore/loader/icon/IconRecord.h
@@ -38,6 +38,8 @@
 #include "PlatformString.h"
 #include "StringHash.h"
 
+#include <time.h>
+
 namespace WebCore { 
 
 class IconDataSnapshot;
diff --git a/src/3rdparty/webkit/WebCore/page/Page.h b/src/3rdparty/webkit/WebCore/page/Page.h
--- a/src/3rdparty/webkit/WebCore/page/Page.h
+++ b/src/3rdparty/webkit/WebCore/page/Page.h
@@ -27,6 +27,8 @@
 #include <wtf/HashSet.h>
 #include <wtf/Noncopyable.h>
 
+#include <time.h>
+
 #if PLATFORM(MAC)
 #include "SchedulePair.h"
 #endif
diff --git a/src/3rdparty/webkit/WebCore/platform/network/ResourceResponseBase.h b/src/3rdparty/webkit/WebCore/platform/network/ResourceResponseBase.h
--- a/src/3rdparty/webkit/WebCore/platform/network/ResourceResponseBase.h
+++ b/src/3rdparty/webkit/WebCore/platform/network/ResourceResponseBase.h
@@ -30,6 +30,8 @@
 #include "HTTPHeaderMap.h"
 #include "KURL.h"
 
+#include <time.h>
+
 #include <wtf/PassOwnPtr.h>
 
 namespace WebCore {
