diff --git a/src/network/kernel/qhostinfo_unix.cpp b/src/network/kernel/qhostinfo_unix.cpp
--- a/src/network/kernel/qhostinfo_unix.cpp
+++ b/src/network/kernel/qhostinfo_unix.cpp
@@ -94,6 +94,7 @@
 
 static void resolveLibrary()
 {
+#ifndef Q_OS_GENODE
 #ifndef QT_NO_LIBRARY
     QLibrary lib(QLatin1String("resolv"));
     if (!lib.load())
@@ -118,6 +119,7 @@
             local_res_ninit = 0;
     }
 #endif
+#endif
 }
 
 QHostInfo QHostInfoAgent::fromName(const QString &hostName)
