Index: kernel/fiasco/src/kern/arm/thread-arm.cpp
===================================================================
--- kernel/fiasco/src/kern/arm/thread-arm.cpp	(revision 38)
+++ kernel/fiasco/src/kern/arm/thread-arm.cpp	(working copy)
@@ -258,6 +258,10 @@
                 && handle_copro_fault[copro](opcode, ts))
               return;
           }
+
+        /* check for ARM default GDB breakpoint */
+        if (!(ts->psr & Proc::Status_thumb) && opcode == 0xe7ffdefe)
+          ts->pc -= 4;
       }
 
 undef_insn:
