Index: kernel/fiasco/src/kern/arm/bsp/realview/timer-arm-mptimer-realview.cpp
===================================================================
--- kernel/fiasco/src/kern/arm/bsp/realview/timer-arm-mptimer-realview.cpp	(revision 38)
+++ kernel/fiasco/src/kern/arm/bsp/realview/timer-arm-mptimer-realview.cpp	(working copy)
@@ -4,7 +4,7 @@
 EXTENSION class Timer
 {
 private:
-  enum { Interval = 104999, /* assumed 210MHz */};
+  enum { Interval = 209999, /* assumed 210MHz */};
 };
 
 // --------------------------------------------------------------------------
@@ -13,5 +13,5 @@
 EXTENSION class Timer
 {
 private:
-  enum { Interval = 49999, };
+  enum { Interval = 99999, /* assumed 100MHz */};
 };
