Index: kernel/fiasco/src/kern/arm/bsp/realview/board_check-arm-realview.cpp
===================================================================
--- kernel/fiasco/src/kern/arm/bsp/realview/board_check-arm-realview.cpp	(revision 38)
+++ kernel/fiasco/src/kern/arm/bsp/realview/board_check-arm-realview.cpp	(working copy)
@@ -46,7 +46,7 @@
 IMPLEMENTATION [arm && realview && realview_vexpress]:
 
 Board_check::id_pair Board_check::ids[] FIASCO_INITDATA = {
-  { 0xffffff00, 0x1190f500 },
+  { 0xcfffff00, 0x0190f500 },
 };
 
 // ------------------------------------------------------------------------
