--- a/src/lib/qemu/hw/usb/bus.c
+++ b/src/lib/qemu/hw/usb/bus.c
@@ -429,13 +429,13 @@
             return;
         }
     } else {
-        if (bus->nfree == 1 && strcmp(object_get_typename(OBJECT(dev)), "usb-hub") != 0) {
-            /* Create a new hub and chain it on */
-            hub = usb_try_new("usb-hub");
-            if (hub) {
-                usb_realize_and_unref(hub, bus, NULL);
-            }
-        }
+        // if (bus->nfree == 1 && strcmp(object_get_typename(OBJECT(dev)), "usb-hub") != 0) {
+        //     /* Create a new hub and chain it on */
+        //     hub = usb_try_new("usb-hub");
+        //     if (hub) {
+        //         usb_realize_and_unref(hub, bus, NULL);
+        //     }
+        // }
         if (bus->nfree == 0) {
             error_setg(errp, "tried to attach usb device %s to a bus "
                        "with no free ports", dev->product_desc);
