--- a/src/lib/qemu/hw/usb/hcd-xhci-pci.c
+++ b/src/lib/qemu/hw/usb/hcd-xhci-pci.c
@@ -252,10 +252,10 @@
     .instance_init = qemu_xhci_instance_init,
 };
 
-static void xhci_register_types(void)
+static void xhci_pci_register_types(void)
 {
     type_register_static(&xhci_pci_info);
     type_register_static(&qemu_xhci_info);
 }
 
-type_init(xhci_register_types)
+type_init(xhci_pci_register_types)
