mouse.patch

diff --git a/src/app/virtualbox/src/VBox/Devices/Input/DrvMouseQueue.cpp b/src/app/virtualbox/src/VBox/Devices/Input/DrvMouseQueue.cpp
index 9d0a8d2..7eb6a09 100644
--- a/src/app/virtualbox/src/VBox/Devices/Input/DrvMouseQueue.cpp
+++ b/src/app/virtualbox/src/VBox/Devices/Input/DrvMouseQueue.cpp
@@ -119,6 +119,7 @@ static DECLCALLBACK(void *)  drvMouseQueueQueryInterface(PPDMIBASE pInterface, c
 /**
  * @interface_method_impl{PDMIMOUSEPORT,pfnPutEvent}
  */
+#include <base/log.h>
 static DECLCALLBACK(int) drvMouseQueuePutEvent(PPDMIMOUSEPORT pInterface,
                                                int32_t dx, int32_t dy,
                                                int32_t dz, int32_t dw,
@@ -141,7 +142,8 @@ static DECLCALLBACK(int) drvMouseQueuePutEvent(PPDMIMOUSEPORT pInterface,
         PDMQueueInsert(pDrv->pQueue, &pItem->Core);
         return VINF_SUCCESS;
     }
-    return VERR_PDM_NO_QUEUE_ITEMS;
+    Genode::error(__func__, " - drop event");
+    return VINF_SUCCESS;
 }
 
 /**
@@ -169,7 +171,8 @@ static DECLCALLBACK(int) drvMouseQueuePutEventAbs(PPDMIMOUSEPORT pInterface,
         PDMQueueInsert(pDrv->pQueue, &pItem->Core);
         return VINF_SUCCESS;
     }
-    return VERR_PDM_NO_QUEUE_ITEMS;
+    Genode::error(__func__, " - drop event");
+    return VINF_SUCCESS;
 }
 
 
