+++ src/app/virtualbox/src/VBox/VMM/VMMAll/PDMAllQueue.cpp
@@ -106,7 +106,7 @@
 # ifdef VBOX_WITH_REM
         REMR3NotifyQueuePending(pVM); /** @todo r=bird: we can remove REMR3NotifyQueuePending and let VMR3NotifyFF do the work. */
 # endif
-        VMR3NotifyGlobalFFU(pVM->pUVM, VMNOTIFYFF_FLAGS_DONE_REM);
+        VMR3NotifyGlobalFFU(pVM->pUVM, VMNOTIFYFF_FLAGS_DONE_REM | VMNOTIFYFF_FLAGS_POKE);
 #endif
     }
     STAM_REL_COUNTER_INC(&pQueue->StatInsert);
