powerbutton.patch

diff --git a/src/app/virtualbox/src/VBox/Main/src-client/ConsoleImpl.cpp b/src/app/virtualbox/src/VBox/Main/src-client/ConsoleImpl.cpp
index b0b04ad..3680be4 100644
--- a/src/app/virtualbox/src/VBox/Main/src-client/ConsoleImpl.cpp
+++ b/src/app/virtualbox/src/VBox/Main/src-client/ConsoleImpl.cpp
@@ -106,7 +106,9 @@
 #if 0
 #include <VBox/vmm/vmapi.h>
 #include <VBox/vmm/vmm.h>
+#endif
 #include <VBox/vmm/pdmapi.h>
+#if 0
 #include <VBox/vmm/pdmasynccompletion.h>
 #include <VBox/vmm/pdmnetifs.h>
 #endif
@@ -2625,6 +2627,7 @@ STDMETHODIMP Console::Resume()
     LogFlowThisFuncLeave();
     return rc;
 }
+#endif
 
 STDMETHODIMP Console::PowerButton()
 {
@@ -2671,6 +2674,7 @@ STDMETHODIMP Console::PowerButton()
     return rc;
 }
 
+#if 0
 STDMETHODIMP Console::GetPowerButtonHandled(BOOL *aHandled)
 {
     LogFlowThisFuncEnter();
