serial.patch

diff --git a/src/app/virtualbox/src/VBox/Devices/Serial/DrvHostSerial.cpp b/src/app/virtualbox/src/VBox/Devices/Serial/DrvHostSerial.cpp
index 0c44fad..34b7fea 100644
--- a/src/app/virtualbox/src/VBox/Devices/Serial/DrvHostSerial.cpp
+++ b/src/app/virtualbox/src/VBox/Devices/Serial/DrvHostSerial.cpp
@@ -903,6 +903,7 @@ static DECLCALLBACK(int) drvHostSerialMonitorThread(PPDMDRVINS pDrvIns, PPDMTHRE
             PDMDrvHlpVMSetRuntimeError(pDrvIns, 0 /*fFlags*/, "DrvHostSerialFail",
                                        N_("Ioctl failed for serial host device '%s' (%Rrc). The device will not work properly"),
                                        pThis->pszDevicePath, RTErrConvertFromErrno(errno));
+            pThread->enmState = PDMTHREADSTATE_TERMINATING;
             break;
         }
 
