sharedfolder_pagelist.patch

diff --git a/src/app/virtualbox/src/VBox/HostServices/SharedFolders/service.cpp b/src/app/virtualbox/src/VBox/HostServices/SharedFolders/service.cpp
index 49a0006..499d541 100644
--- a/src/app/virtualbox/src/VBox/HostServices/SharedFolders/service.cpp
+++ b/src/app/virtualbox/src/VBox/HostServices/SharedFolders/service.cpp
@@ -549,6 +549,14 @@ static DECLCALLBACK(void) svcCall (void *, VBOXHGCMCALLHANDLE callHandle, uint32
                 }
                 else
                 {
+
+
+                /* WORKAROUND start */
+                uint32_t read = 0;
+
+                while (paParms[3].u.uint32 && read < paParms[3].u.uint32) {
+
+
                     /* Execute the function. */
                     if (pStatusLed)
                     {
@@ -562,13 +570,25 @@ static DECLCALLBACK(void) svcCall (void *, VBOXHGCMCALLHANDLE callHandle, uint32
 
                     if (RT_SUCCESS(rc))
                     {
+                        read    += count; 
+                        offset  += count;
+                        pBuffer += count;
+                        count    = paParms[3].u.uint32 - read;
+
                         /* Update parameters.*/
-                        paParms[3].u.uint32 = count;
+                        //paParms[3].u.uint32 = count;
                     }
                     else
                     {
                         paParms[3].u.uint32 = 0;   /* nothing read */
                     }
+
+
+
+                } /* WORKAROUND - End */
+
+
+
                 }
             }
             break;
@@ -616,6 +636,16 @@ static DECLCALLBACK(void) svcCall (void *, VBOXHGCMCALLHANDLE callHandle, uint32
                 }
                 else
                 {
+
+
+
+                /* WORKAROUND start */
+                uint32_t write = 0;
+
+                while (paParms[3].u.uint32 && write < paParms[3].u.uint32) {
+
+
+
                     /* Execute the function. */
                     if (pStatusLed)
                     {
@@ -629,13 +659,25 @@ static DECLCALLBACK(void) svcCall (void *, VBOXHGCMCALLHANDLE callHandle, uint32
 
                     if (RT_SUCCESS(rc))
                     {
+                        write   += count; 
+                        offset  += count;
+                        pBuffer += count;
+                        count    = paParms[3].u.uint32 - write;
+
                         /* Update parameters.*/
-                        paParms[3].u.uint32 = count;
+                        //paParms[3].u.uint32 = count;
                     }
                     else
                     {
                         paParms[3].u.uint32 = 0;   /* nothing read */
                     }
+
+
+
+                } /* WORKAROUND - End */
+
+
+
                 }
             }
             break;
