time-log-deadlock.patch

diff --git a/src/app/virtualbox/src/VBox/Runtime/common/log/log.cpp b/src/app/virtualbox/src/VBox/Runtime/common/log/log.cpp
index 62463c4..f10baad 100644
--- a/src/app/virtualbox/src/VBox/Runtime/common/log/log.cpp
+++ b/src/app/virtualbox/src/VBox/Runtime/common/log/log.cpp
@@ -3128,7 +3128,8 @@ static DECLCALLBACK(size_t) rtLogOutputPrefixed(void *pv, const char *pachChars,
                 {
 
 #if defined(IN_RING3) || defined(IN_RC)
-                    uint64_t u64 = RTTimeProgramMicroTS();
+//                    uint64_t u64 = RTTimeProgramMicroTS();
+                    uint64_t u64 = 0;
                     psz += RTStrFormatNumber(psz, (uint32_t)(u64 / RT_US_1HOUR), 10, 2, 0, RTSTR_F_ZEROPAD);
                     *psz++ = ':';
                     uint32_t u32 = (uint32_t)(u64 % RT_US_1HOUR);
