--- a/src/app/virtualbox/src/VBox/Devices/build/VBoxDD.h
+++ b/src/app/virtualbox/src/VBox/Devices/build/VBoxDD.h
@@ -58,6 +58,7 @@
 extern const PDMDEVREG g_DeviceAudioSniffer;
 extern const PDMDEVREG g_DeviceOHCI;
 extern const PDMDEVREG g_DeviceEHCI;
+extern const PDMDEVREG g_DeviceXHCI;
 extern const PDMDEVREG g_DeviceACPI;
 extern const PDMDEVREG g_DeviceDMA;
 extern const PDMDEVREG g_DeviceFloppyController;
