vmdk.patch

diff --git a/src/app/virtualbox/src/VBox/Storage/VMDK.cpp b/src/app/virtualbox/src/VBox/Storage/VMDK.cpp
index f749ee4..0c5d49e 100644
--- a/src/app/virtualbox/src/VBox/Storage/VMDK.cpp
+++ b/src/app/virtualbox/src/VBox/Storage/VMDK.cpp
@@ -5771,7 +5771,6 @@ static int vmdkRead(void *pBackendData, uint64_t uOffset, size_t cbToRead,
                         }
                         pExtent->uGrainSectorAbs = uSectorExtentAbs;
                         pExtent->uGrain = uSectorExtentRel / pExtent->cSectorsPerGrain;
-                        Assert(uLBA == uSectorExtentRel);
                     }
                     vdIfIoIntIoCtxCopyTo(pImage->pIfIo, pIoCtx,
                                            (uint8_t *)pExtent->pvGrain
@@ -5920,7 +5919,6 @@ static int vmdkWrite(void *pBackendData, uint64_t uOffset, size_t cbToWrite,
                      * which is somewhere between expensive and impossible. */
                     rc = VERR_VD_VMDK_INVALID_STATE;
                     pExtent->uGrainSectorAbs = 0;
-                    AssertRC(rc);
                 }
                 else
                 {
