--- a/alis/a-except.ali
+++ b/alis/a-except.ali
@@ -9,7 +9,7 @@ RV NO_DYNAMIC_SIZED_OBJECTS
 RV NO_IMPLEMENTATION_PRAGMAS
 RV SPARK_05
 
-U ada.exceptions%b	a-except.adb		6bb20b81 NE OL PK
+U ada.exceptions%b	a-except.adb		3813d015 NE OL PK
 W ada%s			ada.ads			ada.ali
 W ada.exceptions%s	a-except.adb		a-except.ali
 W ada_exceptions%s	ada_exceptions.ads	ada_exceptions.ali
@@ -21,7 +21,7 @@ W system.secondary_stack%s  s-secsta.adb	s-secsta.ali
 W system.soft_links%s	s-soflin.adb		s-soflin.ali
 W system.standard_library%s  s-stalib.adb	s-stalib.ali
 
-U ada.exceptions%s	a-except.ads		2c291471 NE OL PR PK
+U ada.exceptions%s	a-except.ads		7f88cfe5 NE OL PR PK
 W ada%s			ada.ads			ada.ali
 W system%s		system.ads		system.ali
 W system.standard_library%s  s-stalib.adb	s-stalib.ali
@@ -39,7 +39,7 @@ D s-imgint.ads		20190214095141 02dbe0c2 system.img_int%s
 D s-parame.ads		20190214095141 82b116e0 system.parameters%s
 D s-secsta.ads		20190214095141 416b2b4a system.secondary_stack%s
 D s-soflin.ads		20190214095141 feb01b63 system.soft_links%s
-D s-stalib.ads		20190214095141 aab0a2bf system.standard_library%s
+D s-stalib.ads		20190214095141 f911792b system.standard_library%s
 D s-stoele.ads		20190214095141 2dc34a04 system.storage_elements%s
 D s-unstyp.ads		20190214095141 9d58a4c0 system.unsigned_types%s
 X 1 ada.ads
@@ -335,6 +335,6 @@ X 9 system.ads
 . 625r67 632r67 641r65 648r65 657r58 664r58 673r54 680r54 689r58 696r58 705r56
 . 712r56
 X 14 s-stalib.ads
-13K16*Standard_Library 2|13w13 51r36 3|16w13 16r42 14|57e28
+13K16*Standard_Library 2|13w13 51r36 3|16w13 16r42 14|68e28
 35P9*Exception_Data_Ptr(34R9) 2|51r53
 
diff --git a/alis/s-soflin.ali b/alis/s-soflin.ali
index 07fc36b..f6f5cac 100644
--- a/alis/s-soflin.ali
+++ b/alis/s-soflin.ali
@@ -27,7 +27,7 @@ D s-parame.ads		20190214095141 82b116e0 system.parameters%s
 D s-secsta.ads		20190214095141 416b2b4a system.secondary_stack%s
 D s-soflin.ads		20190214095141 feb01b63 system.soft_links%s
 D s-soflin.adb		20190214095141 9f3d2c20 system.soft_links%b
-D s-stalib.ads		20190214095141 aab0a2bf system.standard_library%s
+D s-stalib.ads		20190214095141 f911792b system.standard_library%s
 D s-stoele.ads		20190214095141 2dc34a04 system.storage_elements%s
 D s-unstyp.ads		20190214095141 9d58a4c0 system.unsigned_types%s
 X 1 ada.ads
diff --git a/alis/s-stalib.ali b/alis/s-stalib.ali
index 968248d..b9e6d3b 100644
--- a/alis/s-stalib.ali
+++ b/alis/s-stalib.ali
@@ -8,12 +8,13 @@ RV NO_STANDARD_STORAGE_POOLS
 RV NO_UNCHECKED_CONVERSION
 RV NO_DYNAMIC_SIZED_OBJECTS
 RV NO_IMPLEMENTATION_PRAGMAS
+RV NO_OBSOLESCENT_FEATURES
 RV SPARK_05
 
-U system.standard_library%b  s-stalib.adb	622c401b NE OL PK
+U system.standard_library%b  s-stalib.adb	318d9b8f NE OL PK
 W system%s		system.ads		system.ali
 
-U system.standard_library%s  s-stalib.ads	d244d5ce EB NE OL PR PK
+U system.standard_library%s  s-stalib.ads	81e50e5a EB NE OL PR PK
 W ada%s			ada.ads			ada.ali
 W ada.unchecked_conversion%s
 W system%s		system.ads		system.ali
@@ -25,7 +26,7 @@ D ss_utils.ads		20190214095141 f7bf9251 ss_utils%s
 D system.ads		20190214095141 811ff877 system%s
 D s-parame.ads		20190214095141 82b116e0 system.parameters%s
 D s-secsta.ads		20190214095141 416b2b4a system.secondary_stack%s
-D s-stalib.ads		20190214095141 aab0a2bf system.standard_library%s
+D s-stalib.ads		20190214095141 f911792b system.standard_library%s
 D s-stalib.adb		20190214095141 c89ce2a4 system.standard_library%b
 D s-stoele.ads		20190214095141 2dc34a04 system.storage_elements%s
 X 1 ada.ads
@@ -33,24 +34,28 @@ X 1 ada.ads
 X 2 a-unccon.ads
 20v14*Unchecked_Conversion 7|11w10 32r14
 X 4 system.ads
-47K9*System 178e11 7|13r9 32r36 42r31 44r31 57r5 8|1r14 7r5
+47K9*System 178e11 7|13r9 32r36 42r31 44r31 68r5 8|1r14 7r5
 77M9*Address 7|32r43 42r38 44r38
+82m4*Null_Address{77M9} 7|55r32
 X 7 s-stalib.ads
-13K16*Standard_Library 4|47k9 7|57l12 57e28 8|1b21 7l12 7t28
+13K16*Standard_Library 4|47k9 7|68l12 68e28 8|1b21 7l12 7t28
 17A12*Big_String{string}<integer> 18r36 24r38
 24P9*Big_String_Ptr(17A12) 25r8 32r52
 31V13*To_Ptr[2|20]{24P9}
-34R9*Exception_Data 35r42 38c9 46e14
+34R9*Exception_Data 35r42 38c9 46e14 49r35
 35P9*Exception_Data_Ptr(34R9) 43r31
 36P9*Raise_Action 45r31
-39b7*Not_Handled_By_Others{boolean}
-40e7*Lang{character}
-41i7*Name_Length{natural}
-42m7*Full_Name{4|77M9}
-43p7*HTable_Ptr{35P9}
-44m7*Foreign_Data{4|77M9}
-45p7*Raise_Hook{36P9}
-48U14*Adafinal
-55V13*Dummy{string} 55>20
-55a20 S{string} 55r50
+38R9*Exception_Data
+39b7*Not_Handled_By_Others{boolean} 50m7
+40e7*Lang{character} 51m7
+41i7*Name_Length{natural} 52m7
+42m7*Full_Name{4|77M9} 53m7
+43p7*HTable_Ptr{35P9} 54m7
+44m7*Foreign_Data{4|77M9} 55m7
+45p7*Raise_Hook{36P9} 56m7
+48a4*Constraint_Error_Name{string} 52r32 53r32
+49r4*Constraint_Error_Def{38R9} 57m22 57r22
+59U14*Adafinal
+66V13*Dummy{string} 66>20
+66a20 S{string} 66r50
 
