diff -urN lwip-1.3.2.orig/src/include/lwip/arch.h lwip-1.3.2/src/include/lwip/arch.h
--- lwip-1.3.2.orig/src/include/lwip/arch.h	2009-12-24 16:38:19.000000000 +0100
+++ lwip-1.3.2/src/include/lwip/arch.h	2010-11-01 15:43:34.000000000 +0100
@@ -224,6 +224,12 @@
 extern int errno;
 #endif
 
+#else
+
+#include <errno.h>
+/* errno codes which are not defined in Genode libc's errno.h */
+#define ENSRNOTFOUND 163  /* Domain name not found */
+
 #endif /* LWIP_PROVIDE_ERRNO */
 
 #ifdef __cplusplus
