--- /tmp/freebsd-libc.cvs/stdlib/malloc.c	2008-10-21 15:50:04.000000000 +0100
+++ libc/stdlib/malloc.c	2008-10-22 18:08:40.000000000 +0100
@@ -153,7 +153,7 @@
  * unnecessary, but we are burdened by history and the lack of resource limits
  * for anonymous mapped memory.
  */
-#define	MALLOC_DSS
+//#define	MALLOC_DSS
 
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdlib/malloc.c,v 1.179 2008/09/10 14:27:34 jasone Exp $");
