--- src/kernel/sel4/src/arch/arm/kernel/boot.c
+++ src/kernel/sel4/src/arch/arm/kernel/boot.c
@@ -582,6 +582,14 @@
         dtb_end_p = dtb_addr_p + dtb_size;
     }
 
+    /*
+     * no DTB support on Genode/seL4 - avoids assertion in
+     *
+     * ERROR: DTB at [defffaff..defffb00] exceeds PADDR_TOP (2ff00000)
+     * src/kernel/sel4/src/arch/arm/kernel/boot.c:668 in function init_kernel, saying "ERROR: kernel init failed"
+     */
+    dtb_addr_p = dtb_size = 0;
+
 #ifdef ENABLE_SMP_SUPPORT
     /* we assume there exists a cpu with id 0 and will use it for bootstrapping */
     if (getCurrentCPUIndex() == 0) {
