--- src/kernel/sel4/include/plat/pc99/plat/64/plat_mode/machine/hardware.h
+++ src/kernel/sel4/include/plat/pc99/plat/64/plat_mode/machine/hardware.h
@@ -87,7 +87,7 @@
 #define PPTR_TOP UL_CONST(0xffffffff80000000)
 
 /* The physical memory address to use for mapping the kernel ELF */
-#define KERNEL_ELF_PADDR_BASE UL_CONST(0x00100000)
+#define KERNEL_ELF_PADDR_BASE UL_CONST(0x00200000)
 
 /* Kernel mapping starts directly after the physical memory window */
 #define KERNEL_ELF_BASE (PPTR_TOP + KERNEL_ELF_PADDR_BASE)
--- src/kernel/sel4/include/plat/pc99/plat/32/plat_mode/machine/hardware.h
+++ src/kernel/sel4/include/plat/pc99/plat/32/plat_mode/machine/hardware.h
@@ -74,7 +74,7 @@
 #define PPTR_TOP KS_LOG_PPTR
 
 /* The physical memory address to use for mapping the kernel ELF */
-#define KERNEL_ELF_PADDR_BASE UL_CONST(0x00100000)
+#define KERNEL_ELF_PADDR_BASE UL_CONST(0x00200000)
 
 /* The base address in virtual memory to use for the kernel ELF mapping */
 #define KERNEL_ELF_BASE (PPTR_BASE + KERNEL_ELF_PADDR_BASE)
