--- a/src/lib/mesa/src/gallium/drivers/iris/iris_bufmgr.c
+++ b/src/lib/mesa/src/gallium/drivers/iris/iris_bufmgr.c
@@ -101,6 +101,8 @@
 
 #define FILE_DEBUG_FLAG DEBUG_BUFMGR
 
+void drm_unmap_ppgtt(__u32 handle);
+
 static inline int
 atomic_add_unless(int *v, int add, int unless)
 {
@@ -439,6 +441,9 @@
     */
    if (memzone != iris_memzone_for_address(bo->gtt_offset) ||
        bo->gtt_offset % alignment != 0) {
+
+
+      drm_unmap_ppgtt(bo->gem_handle);
       vma_free(bufmgr, bo->gtt_offset, bo->size);
       bo->gtt_offset = 0ull;
    }
