--- a/src/lib/mesa/src/gallium/drivers/iris/iris_bufmgr.c
+++ b/src/lib/mesa/src/gallium/drivers/iris/iris_bufmgr.c
@@ -1297,6 +1297,8 @@
    return ret;
 }
 
+struct iris_bufmgr * single_bufmgr = NULL;
+
 static void
 iris_bufmgr_destroy(struct iris_bufmgr *bufmgr)
 {
@@ -1814,10 +1816,13 @@
 static struct iris_bufmgr *
 iris_bufmgr_create(struct gen_device_info *devinfo, int fd, bool bo_reuse)
 {
-   uint64_t gtt_size = iris_gtt_size(fd);
+printf("%s:%u\n", __func__, __LINE__);
+//   uint64_t gtt_size = iris_gtt_size(fd);
+   uint64_t gtt_size = 2 * IRIS_MEMZONE_OTHER_START;
    if (gtt_size <= IRIS_MEMZONE_OTHER_START)
       return NULL;
 
+printf("%s:%u\n", __func__, __LINE__);
    struct iris_bufmgr *bufmgr = calloc(1, sizeof(*bufmgr));
    if (bufmgr == NULL)
       return NULL;
@@ -1920,14 +1925,18 @@
 struct iris_bufmgr *
 iris_bufmgr_get_for_fd(struct gen_device_info *devinfo, int fd, bool bo_reuse)
 {
+printf("%s:%u\n", __func__, __LINE__);
+#if 0
    struct stat st;
 
    if (fstat(fd, &st))
       return NULL;
+#endif
 
    struct iris_bufmgr *bufmgr = NULL;
 
    mtx_lock(&global_bufmgr_list_mutex);
+#if 0
    list_for_each_entry(struct iris_bufmgr, iter_bufmgr, &global_bufmgr_list, link) {
       struct stat iter_st;
       if (fstat(iter_bufmgr->fd, &iter_st))
@@ -1939,7 +1948,13 @@
          goto unlock;
       }
    }
+#endif
 
+   if (single_bufmgr) {
+     struct iris_bufmgr * iter_bufmgr = single_bufmgr;
+     bufmgr = iris_bufmgr_ref(iter_bufmgr);
+     goto unlock;
+   } else
    bufmgr = iris_bufmgr_create(devinfo, fd, bo_reuse);
    if (bufmgr)
       list_addtail(&bufmgr->link, &global_bufmgr_list);
