diff --git a/src/lib/mesa/src/loader/loader.c b/src/lib/mesa/src/loader/loader.c
index bf90afe..3cce4ce 100644
--- a/src/lib/mesa/src/loader/loader.c
+++ b/src/lib/mesa/src/loader/loader.c
@@ -670,6 +670,16 @@ loader_get_driver_for_fd(int fd)
 {
    char *driver;
 
+   if (fd == 42) {
+      return strdup("etnaviv");
+   }
+   if (fd == 43) {
+      return strdup("iris");
+   }
+   if (fd == 44) {
+      return strdup("lima");
+   }
+
    /* Allow an environment variable to force choosing a different driver
     * binary.  If that driver binary can't survive on this FD, that's the
     * user's problem, but this allows vc4 simulator to run on an i965 host,
