diff --git a/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.c b/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.c
index 992f0e3..1f682fc 100644
--- a/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.c
+++ b/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.c
@@ -1170,6 +1170,9 @@ dri2_initialize(_EGLDisplay *disp)
    case _EGL_PLATFORM_ANDROID:
       ret = dri2_initialize_android(disp);
       break;
+   case _EGL_PLATFORM_GENODE:
+      ret = dri2_initialize_genode(disp);
+      break;
    default:
       unreachable("Callers ensure we cannot get here.");
       return EGL_FALSE;
diff --git a/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.h b/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.h
index 57d8820..0d1be58 100644
--- a/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.h
+++ b/src/lib/mesa/src/egl/drivers/dri2/egl_dri2.h
@@ -356,6 +356,13 @@ struct dri2_egl_surface {
    bool received_dmabuf_feedback;
 #endif
 
+#ifdef HAVE_GENODE_PLATFORM
+   struct Genode_egl_window *g_win;
+   __DRIbuffer              *dri_buffers[__DRI_BUFFER_COUNT];
+   __DRIimage               *back_image;
+#endif
+
+
 #ifdef HAVE_DRM_PLATFORM
    struct gbm_dri_surface *gbm_surf;
 #endif
@@ -606,6 +613,9 @@ dri2_initialize_android(_EGLDisplay *disp)
 EGLBoolean
 dri2_initialize_surfaceless(_EGLDisplay *disp);
 
+EGLBoolean
+dri2_initialize_genode(_EGLDisplay *disp);
+
 EGLBoolean
 dri2_initialize_device(_EGLDisplay *disp);
 static inline void
diff --git a/src/lib/mesa/src/egl/main/egldisplay.c b/src/lib/mesa/src/egl/main/egldisplay.c
index 4eadb8f..6a46d79 100644
--- a/src/lib/mesa/src/egl/main/egldisplay.c
+++ b/src/lib/mesa/src/egl/main/egldisplay.c
@@ -83,6 +83,7 @@ static const struct {
    {_EGL_PLATFORM_SURFACELESS, "surfaceless"},
    {_EGL_PLATFORM_DEVICE, "device"},
    {_EGL_PLATFORM_WINDOWS, "windows"},
+   {_EGL_PLATFORM_GENODE, "genode"},
 };
 
 /**
diff --git a/src/lib/mesa/src/egl/main/egldisplay.h b/src/lib/mesa/src/egl/main/egldisplay.h
index b3510ae..14e8354 100644
--- a/src/lib/mesa/src/egl/main/egldisplay.h
+++ b/src/lib/mesa/src/egl/main/egldisplay.h
@@ -55,6 +55,7 @@ enum _egl_platform_type {
    _EGL_PLATFORM_SURFACELESS,
    _EGL_PLATFORM_DEVICE,
    _EGL_PLATFORM_WINDOWS,
+   _EGL_PLATFORM_GENODE,
 
    _EGL_NUM_PLATFORMS,
    _EGL_INVALID_PLATFORM = -1
diff --git a/src/lib/mesa/src/egl/main/egllog.c b/src/lib/mesa/src/egl/main/egllog.c
index 678bb75..9c16c60 100644
--- a/src/lib/mesa/src/egl/main/egllog.c
+++ b/src/lib/mesa/src/egl/main/egllog.c
@@ -55,7 +55,7 @@
 #endif /* HAVE_ANDROID_PLATFORM */
 
 #define MAXSTRING          1000
-#define FALLBACK_LOG_LEVEL _EGL_WARNING
+#define FALLBACK_LOG_LEVEL _EGL_WARNING
 
 static struct {
    simple_mtx_t mutex;
diff --git a/src/lib/mesa/src/loader/loader.c b/src/lib/mesa/src/loader/loader.c
index 19c456c..bf90afe 100644
--- a/src/lib/mesa/src/loader/loader.c
+++ b/src/lib/mesa/src/loader/loader.c
@@ -800,8 +800,7 @@ loader_open_driver_lib(const char *driver_name,
          next = end;
 
       len = next - p;
-      snprintf(path, sizeof(path), "%.*s/tls/%s%s.so", len,
-               p, driver_name, lib_suffix);
+      snprintf(path, sizeof(path), "%.*s/mesa_gpu.lib.so", len, p);
       driver = dlopen(path, RTLD_NOW | RTLD_GLOBAL);
       if (driver == NULL) {
          snprintf(path, sizeof(path), "%.*s/%s%s.so", len,
diff --git a/src/lib/mesa/src/util/u_thread.c b/src/lib/mesa/src/util/u_thread.c
index c508733..b186182 100644
--- a/src/lib/mesa/src/util/u_thread.c
+++ b/src/lib/mesa/src/util/u_thread.c
@@ -154,7 +154,8 @@ util_set_thread_affinity(thrd_t thread,
 int64_t
 util_thread_get_time_nano(thrd_t thread)
 {
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__) \
+    && !defined(__GENODE__)
    struct timespec ts;
    clockid_t cid;
 
@@ -174,7 +175,8 @@ util_thread_get_time_nano(thrd_t thread)
 #endif
 }
 
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) \
+    && !defined(__GENODE__)
 
 void util_barrier_init(util_barrier *barrier, unsigned count)
 {
diff --git a/src/lib/mesa/src/util/u_thread.h b/src/lib/mesa/src/util/u_thread.h
index b5c7667..7065bc8 100644
--- a/src/lib/mesa/src/util/u_thread.h
+++ b/src/lib/mesa/src/util/u_thread.h
@@ -116,7 +116,8 @@ static inline bool u_thread_is_self(thrd_t thread)
  * util_barrier
  */
 
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) \
+    && !defined(__GENODE__)
 
 typedef pthread_barrier_t util_barrier;
 
