example_virtualkeyboard.patch

diff --git a/qtvirtualkeyboard/examples/virtualkeyboard/basic/basic.pro b/qtvirtualkeyboard/examples/virtualkeyboard/basic/basic.pro
index 27ad414..6383ffe 100644
--- a/qtvirtualkeyboard/examples/virtualkeyboard/basic/basic.pro
+++ b/qtvirtualkeyboard/examples/virtualkeyboard/basic/basic.pro
@@ -3,10 +3,10 @@ TARGET = basic
 QT += qml quick
 SOURCES += main.cpp
 CONFIG += link_pkgconfig
-static {
-    QT += svg
-    QTPLUGIN += qtvirtualkeyboardplugin
-}
+#static {
+QT += svg
+QTPLUGIN += qtvirtualkeyboardplugin
+#}
 
 target.path = $$[QT_INSTALL_EXAMPLES]/virtualkeyboard/basic
 INSTALLS += target
@@ -22,13 +22,13 @@ OTHER_FILES += \
     content/TextArea.qml \
     content/TextField.qml \
 
-disable-xcb {
-    message("The disable-xcb option has been deprecated. Please use disable-desktop instead.")
-    CONFIG += disable-desktop
-}
+#disable-xcb {
+#    message("The disable-xcb option has been deprecated. Please use disable-desktop instead.")
+#    CONFIG += disable-desktop
+#}
 
-disable-desktop|android-embedded|!isEmpty(CROSS_COMPILE)|qnx {
-    DEFINES += MAIN_QML=\\\"basic-b2qt.qml\\\"
-} else {
-    DEFINES += MAIN_QML=\\\"Basic.qml\\\"
-}
+#disable-desktop|android-embedded|!isEmpty(CROSS_COMPILE)|qnx {
+#    DEFINES += MAIN_QML=\\\"basic-b2qt.qml\\\"
+#} else {
+#    DEFINES += MAIN_QML=\\\"Basic.qml\\\"
+#}
