qtbase_fixes.patch

diff --git a/qtbase/src/widgets/kernel/qwidgetwindow.cpp b/qtbase/src/widgets/kernel/qwidgetwindow.cpp
index fbc71cd..143cc60 100644
--- a/qtbase/src/widgets/kernel/qwidgetwindow.cpp
+++ b/qtbase/src/widgets/kernel/qwidgetwindow.cpp
@@ -369,6 +369,9 @@ bool QWidgetWindow::event(QEvent *event)
         m_widget->repaint();
         return true;
 
+    case QEvent::MetaCall:
+    	return QWindow::event(event);
+
     default:
         break;
     }
diff --git a/qtbase/src/widgets/widgets/qwidgetanimator.cpp b/qtbase/src/widgets/widgets/qwidgetanimator.cpp
index 486d65d..6ab6798 100644
--- a/qtbase/src/widgets/widgets/qwidgetanimator.cpp
+++ b/qtbase/src/widgets/widgets/qwidgetanimator.cpp
@@ -81,7 +81,8 @@ void QWidgetAnimator::abort(QWidget *w)
 void QWidgetAnimator::animationFinished()
 {
     QPropertyAnimation *anim = qobject_cast<QPropertyAnimation*>(sender());
-    abort(static_cast<QWidget*>(anim->targetObject()));
+    if (anim)
+    	abort(static_cast<QWidget*>(anim->targetObject()));
 }
 #endif // animation
 
