diff --git a/qtbase/src/widgets/kernel/qwidgetwindow.cpp b/qtbase/src/widgets/kernel/qwidgetwindow.cpp
index b91fbcf..3221bf3 100644
--- a/qtbase/src/widgets/kernel/qwidgetwindow.cpp
+++ b/qtbase/src/widgets/kernel/qwidgetwindow.cpp
@@ -122,6 +122,7 @@ QWidgetWindow::QWidgetWindow(QWidget *widget)
     : QWindow(*new QWidgetWindowPrivate(), 0)
     , m_widget(widget)
 {
+    setTitle(widget->windowTitle());
     updateObjectName();
     // Enable QOpenGLWidget/QQuickWidget children if the platform plugin supports it,
     // and the application developer has not explicitly disabled it.
