qtdeclarative_genode.patch

diff --git a/qtdeclarative/src/qml/jsruntime/qv4engine.cpp b/qtdeclarative/src/qml/jsruntime/qv4engine.cpp
index f5c5c49..1d70dce 100644
--- a/qtdeclarative/src/qml/jsruntime/qv4engine.cpp
+++ b/qtdeclarative/src/qml/jsruntime/qv4engine.cpp
@@ -133,6 +133,10 @@
 
 #endif // #ifndef V4_BOOTSTRAP
 
+#if defined(Q_OS_GENODE)
+#include <base/thread.h>
+#endif
+
 QT_BEGIN_NAMESPACE
 
 Q_LOGGING_CATEGORY(lcTracingAll, "qt.v4.tracing.all")
diff --git a/qtdeclarative/src/qml/jsruntime/qv4global_p.h b/qtdeclarative/src/qml/jsruntime/qv4global_p.h
index d47393b..f0123b3 100644
--- a/qtdeclarative/src/qml/jsruntime/qv4global_p.h
+++ b/qtdeclarative/src/qml/jsruntime/qv4global_p.h
@@ -116,7 +116,7 @@ inline double trunc(double d) { return d > 0 ? floor(d) : ceil(d); }
 
 // Black list some platforms
 #if defined(V4_ENABLE_JIT)
-#if defined(Q_OS_IOS) || defined(Q_OS_TVOS)
+#if defined(Q_OS_IOS) || defined(Q_OS_TVOS) || defined(Q_OS_GENODE)
 #    undef V4_ENABLE_JIT
 #endif
 #endif
diff --git a/qtdeclarative/src/qml/qml/qqmlimport.cpp b/qtdeclarative/src/qml/qml/qqmlimport.cpp
index 5a13644..2b8c225 100644
--- a/qtdeclarative/src/qml/qml/qqmlimport.cpp
+++ b/qtdeclarative/src/qml/qml/qqmlimport.cpp
@@ -1913,6 +1913,9 @@ QString QQmlImportDatabase::resolvePlugin(QQmlTypeLoader *typeLoader,
         QLatin1String(".so"),
         QLatin1String(".bundle")
     };
+#elif defined(Q_OS_GENODE)
+    static const QString prefix = QLatin1String("qt5_");
+    static const QStringList suffixes = { QLatin1String(".lib.so") };
 # else  // Unix
     static const QString prefix = QLatin1String("lib");
     static const QStringList suffixes = { QLatin1String(".so") };
