qtdeclarative_configuration.patch

diff --git a/qtdeclarative/tools/tools.pro b/qtdeclarative/tools/tools.pro
index 73cb6e2..462b197 100644
--- a/qtdeclarative/tools/tools.pro
+++ b/qtdeclarative/tools/tools.pro
@@ -20,8 +20,7 @@ qtConfig(thread):!android|android_app:!wasm {
     qtHaveModule(quick) {
         !static: {
             SUBDIRS += \
-                qmlscene \
-                qmltime
+                qmlscene
 
             qtConfig(regularexpression):qtConfig(process) {
                 SUBDIRS += \
