qttools_configuration.patch

diff --git a/qttools/src/designer/src/lib/lib.pro b/qttools/src/designer/src/lib/lib.pro
index de0dc73..c294782 100644
--- a/qttools/src/designer/src/lib/lib.pro
+++ b/qttools/src/designer/src/lib/lib.pro
@@ -12,8 +12,8 @@ DEFINES += \
 static:DEFINES += QT_DESIGNER_STATIC
 
 include(extension/extension.pri)
-include(sdk/sdk.pri)
-include(shared/shared.pri)
+#include(sdk/sdk.pri)
+#include(shared/shared.pri)
 include(uilib/uilib.pri)
 PRECOMPILED_HEADER=lib_pch.h
 
diff --git a/qttools/src/designer/src/src.pro b/qttools/src/designer/src/src.pro
index 79e9e64..bee94b6 100644
--- a/qttools/src/designer/src/src.pro
+++ b/qttools/src/designer/src/src.pro
@@ -7,8 +7,6 @@ SUBDIRS = \
 qtConfig(process) {
     SUBDIRS += \
         lib \
-        components \
-        designer
 
     lib.depends = uiplugin
     components.depends = lib
diff --git a/qttools/src/src.pro b/qttools/src/src.pro
index 5c256ea..a7edd3b 100644
--- a/qttools/src/src.pro
+++ b/qttools/src/src.pro
@@ -16,7 +16,7 @@ qtHaveModule(widgets) {
     }
 }
 
-SUBDIRS += linguist \
+SUBDIRS += \
     qtattributionsscanner
 
 qtConfig(library) {
