diff --git a/conts/userlibs/SConstruct b/conts/userlibs/SConstruct
index 41c7913..421b563 100644
--- a/conts/userlibs/SConstruct
+++ b/conts/userlibs/SConstruct
@@ -11,6 +11,7 @@ PROJRELROOT = '../..'
 sys.path.append(PROJRELROOT)
 
 from scripts.config.config_invoke import *
+from scripts.config.projpaths import *
 
 config = configuration_retrieve()
 gcc_arch_flag = config.gcc_arch_flag
@@ -28,7 +29,7 @@ env = Environment(CC = config.toolchain_userspace + 'gcc',
                   ASFLAGS = ['-D__ASSEMBLY__', '-march=' + gcc_arch_flag],
                   ENV = {'PATH' : os.environ['PATH']},
                   LIBS = 'gcc',	# libgcc.a - Required for division routines.
-                  CPPPATH = KERNEL_HEADERS,
+                  CPPPATH = [KERNEL_HEADERS, LIBC_INCLUDE],
                   CPPFLAGS = '-include l4/config.h -include l4/macros.h -include l4/types.h')
 
 # Set the build directory for this source tree
