diff --git a/projects/execute b/projects/execute
index 43820f0ed..e2dbda43c 100755
--- a/projects/execute
+++ b/projects/execute
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 #
 # Build system execution wrapper.
 #
diff --git a/tools/scripts/mulog-subject.py b/tools/scripts/mulog-subject.py
index cb6e355dd..51965a797 100755
--- a/tools/scripts/mulog-subject.py
+++ b/tools/scripts/mulog-subject.py
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 from __future__ import print_function
 
diff --git a/tools/scripts/mulog.py b/tools/scripts/mulog.py
index 6d7fe8bb2..671913441 100755
--- a/tools/scripts/mulog.py
+++ b/tools/scripts/mulog.py
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 import sys
 import re

