diff --git a/contrib/ada-bfd/Makefile b/contrib/ada-bfd/Makefile
index cd776781e..36c87b126 100644
--- a/contrib/ada-bfd/Makefile
+++ b/contrib/ada-bfd/Makefile
@@ -1,6 +1,6 @@
-REV = 235a23452f7c5104ce6270a84f9794f497d496aa
+REV = 1ee4c51befbda90360e519f5b76c17db124e2973
 PKG = ada-bfd
-SRC = https://github.com/stcarrez/ada-bfd.git
+SRC = https://github.com/skalk/ada-bfd.git
 
 define CMD_CONFIGURE
 	cd $(WRK) && ./configure --prefix=$(BUILDDIR)
diff --git a/contrib/projects/bfdada.gpr b/contrib/projects/bfdada.gpr
index 8a9cd711a..6cbe14152 100644
--- a/contrib/projects/bfdada.gpr
+++ b/contrib/projects/bfdada.gpr
@@ -1,8 +1,8 @@
 library project Bfdada is
-   for Source_Dirs use ("../build/share/ada/adainclude/bfdada");
+   for Source_Dirs use ("../build/include/bfdada.static");
    for Library_Name use "ada-bfd";
-   for Library_Dir use "../build/lib";
-   for Library_ALI_Dir use "../build/lib/ada/adalib/bfdada";
+   for Library_Dir use "../build/lib/bfdada.static";
+   for Library_ALI_Dir use "../build/lib/bfdada.static";
    for Library_Kind use "static";
    for Externally_Built use "true";
 end Bfdada;
