FOC: enable GCC 12

diff --git a/src/Makeconf b/src/Makeconf
index de5b656..7660daa 100644
--- a/src/Makeconf
+++ b/src/Makeconf
@@ -244,7 +244,7 @@ ifeq ($(CC_TYPE),gcc)
 endif
 
 
-CC_WHITELIST-gcc     := 4.7 4.8 4.9 5 6 7 8 9
+CC_WHITELIST-gcc     := 4.7 4.8 4.9 5 6 7 8 9 10 11 12
 CC_WHITELIST-clang   := 3.6 3.7 3.8 3.9 4.0 5.0 6.0 7.0 8.0 9.0
 CC_WHITELIST         := $(CC_WHITELIST-$(CC_TYPE))
 
