+++ src/lib/sdl/src/video/SDL_video.c
@@ -28,8 +28,8 @@
 #include "SDL_blit.h"
 #include "SDL_pixels_c.h"
 #include "SDL_cursor_c.h"
-#include "../events/SDL_sysevents.h"
-#include "../events/SDL_events_c.h"
+#include "SDL_sysevents.h"
+#include "SDL_events_c.h"
 
 /* Available video drivers */
 static VideoBootStrap *bootstrap[] = {
@@ -120,6 +120,9 @@
 #if SDL_VIDEO_DRIVER_AALIB
 	&AALIB_bootstrap,
 #endif
+#if SDL_VIDEO_DRIVER_GENODE_FB
+	&Genode_fb_bootstrap,
+#endif
 #if SDL_VIDEO_DRIVER_DUMMY
 	&DUMMY_bootstrap,
 #endif
+++ src/lib/sdl/src/video/SDL_sysvideo.h
@@ -404,6 +404,9 @@
 #if SDL_VIDEO_DRIVER_AALIB
 extern VideoBootStrap AALIB_bootstrap;
 #endif
+#if SDL_VIDEO_DRIVER_GENODE_FB
+extern VideoBootStrap Genode_fb_bootstrap;
+#endif
 #if SDL_VIDEO_DRIVER_DUMMY
 extern VideoBootStrap DUMMY_bootstrap;
 #endif
+++ src/lib/sdl/include/SDL_opengl.h
@@ -42,7 +42,7 @@
 #include <glu.h>	/* Header File For The GLU Library */
 #else
 #include <GL/gl.h>	/* Header File For The OpenGL Library */
-#include <GL/glu.h>	/* Header File For The GLU Library */
+//#include <GL/glu.h>	/* Header File For The GLU Library */
 #endif
 #ifndef NO_SDL_GLEXT
 #undef __glext_h_
