diff --git a/julian/model/82576vf.cc b/julian/model/82576vf.cc
index b4d29a5..db8d2f3 100644
--- a/julian/model/82576vf.cc
+++ b/julian/model/82576vf.cc
@@ -950,7 +950,7 @@ public:
 
 };
 
-PARAM_HANDLER(82576vf,
+PARAM_HANDLER(i82576vf,
 	      "82576vf:[promisc][,mem_mmio][,mem_msix][,txpoll_us][,rx_map] - attach an Intel 82576VF to the PCI bus.",
 	      "promisc   - if !=0, be always promiscuous (use for Linux VMs that need it for bridging) (Default 1)",
 	      "txpoll_us - if !=0, map TX registers to guest and poll them every txpoll_us microseconds. (Default 0)",
