+++ src/VBox/Devices/PC/DevACPI.cpp
@@ -3254,13 +3256,13 @@
     } while (0)
     R(SMI_CMD,        1, acpiSmiWrite,          NULL,                "ACPI SMI");
 #ifdef DEBUG_ACPI
-    R(DEBUG_HEX,      1, acpiDhexWrite,         NULL,                "ACPI Debug hex");
-    R(DEBUG_CHR,      1, acpiDchrWrite,         NULL,                "ACPI Debug char");
+    R(DEBUG_HEX,      4, acpiDhexWrite,         NULL,                "ACPI Debug hex");
+//    R(DEBUG_CHR,      1, acpiDchrWrite,         NULL,                "ACPI Debug char");
 #endif
-    R(BAT_INDEX,      1, acpiBatIndexWrite,     NULL,                "ACPI Battery status index");
-    R(BAT_DATA,       1, NULL,                  acpiBatDataRead,     "ACPI Battery status data");
-    R(SYSI_INDEX,     1, acpiSysInfoIndexWrite, NULL,                "ACPI system info index");
-    R(SYSI_DATA,      1, acpiSysInfoDataWrite,  acpiSysInfoDataRead, "ACPI system info data");
+    R(BAT_INDEX,      4, acpiBatIndexWrite,     NULL,                "ACPI Battery status index");
+    R(BAT_DATA,       4, NULL,                  acpiBatDataRead,     "ACPI Battery status data");
+    R(SYSI_INDEX,     4, acpiSysInfoIndexWrite, NULL,                "ACPI system info index");
+    R(SYSI_DATA,      4, acpiSysInfoDataWrite,  acpiSysInfoDataRead, "ACPI system info data");
     R(ACPI_RESET_BLK, 1, acpiResetWrite,        NULL,                "ACPI Reset");
 #undef R
 
