+++ src/app/netperf/src/netlib.c
@@ -844,6 +844,10 @@
     exit(1);
   }
 
+  /* On Genode libc port -1 is returned for sysconf(_SC_NPROCESSORS_ONLN) */
+  if (temp_cpus < 0)
+    temp_cpus = 1;
+  
   return(temp_cpus);
 
 }
@@ -1095,8 +1095,13 @@
 
 #endif /* WIN32 */
 
+#ifdef GENODE_BUILD
+void
+start_do_not_use_timer(int time)
+#else
 void
 start_timer(int time)
+#endif
 {
 
 #ifdef WIN32
@@ -3059,7 +3063,9 @@
 	      (which == SEND_BUFFER) ? "SO_SNDBUF" : "SO_RCVBUF",
 	      errno);
       fflush(where);
+#ifndef GENODE_BUILD
       exit(1);
+#endif
     }
     if (debug > 1) {
       fprintf(where, "netperf: set_sock_buffer: %s of %d requested.\n",
