diff --git a/libsel4/arch_include/ia32/sel4/arch/syscalls.h b/libsel4/arch_include/ia32/sel4/arch/syscalls.h
index aef583b..d550343 100644
--- a/libsel4/arch_include/ia32/sel4/arch/syscalls.h
+++ b/libsel4/arch_include/ia32/sel4/arch/syscalls.h
@@ -475,7 +475,7 @@ seL4_DebugPutChar(char c)
         "sysenter          \n"
         "popl %%ebp        \n"
         :
-        : "a" (seL4_SysDebugPutChar),
+        : "a" ((int)seL4_SysDebugPutChar),
         "b" (c)
         : "%ecx", "%edx", "%esi", "%edi", "memory"
     );
