+++ src/lib/sdl_net/SDL_net.h
@@ -350,6 +350,13 @@
 extern no_parse_DECLSPEC char * SDLCALL SDLNet_GetError(void);
 */
 
+#ifdef __cplusplus
+#define SDL_reinterpret_cast(type, expression) reinterpret_cast<type>(expression)
+#define SDL_static_cast(type, expression) static_cast<type>(expression)
+#else
+#define SDL_reinterpret_cast(type, expression) ((type)(expression))
+#define SDL_static_cast(type, expression) ((type)(expression))
+#endif
 
 /* Inline macro functions to read/write network data */
 
