eminternal.patch

diff --git a/src/app/virtualbox/src/VBox/VMM/include/EMInternal.h b/src/app/virtualbox/src/VBox/VMM/include/EMInternal.h
index 11a7ac3..02e6e29 100644
--- a/src/app/virtualbox/src/VBox/VMM/include/EMInternal.h
+++ b/src/app/virtualbox/src/VBox/VMM/include/EMInternal.h
@@ -361,6 +361,8 @@ typedef struct EMCPU
 #ifdef VBOX_WITH_RAW_MODE
     /** Pointer to the PATM status structure. (R3 Ptr) */
     R3PTRTYPE(PPATMGCSTATE) pPatmGCState;
+#else
+    RTR3PTR dummyDONTUSE;
 #endif
 
     /** Pointer to the guest CPUM state. (R3 Ptr) */
