--- a/src/app/virtualbox/src/VBox/VMM/VMMR3/EM.cpp
+++ b/src/app/virtualbox/src/VBox/VMM/VMMR3/EM.cpp
@@ -2211,10 +2227,11 @@
                  */
                 case VINF_EM_RESCHEDULE_REM:
                     Assert(!pVM->em.s.fIemExecutesAll || pVCpu->em.s.enmState != EMSTATE_IEM);
-                    if (HMIsEnabled(pVM))
+                    if (HMIsEnabled(pVM) &&
+                        !(VM_FF_IS_PENDING(pVM, VM_FF_ALL_REM_MASK) || VMCPU_FF_IS_PENDING(pVCpu, VMCPU_FF_ALL_REM_MASK)))
                     {
-                        Log2(("EMR3ExecuteVM: VINF_EM_RESCHEDULE_REM: %d -> %d (EMSTATE_IEM_THEN_REM)\n",
-                              enmOldState, EMSTATE_IEM_THEN_REM));
+                        Log2(("EMR3ExecuteVM: VINF_EM_RESCHEDULE_REM: %d, %d -> %d (EMSTATE_IEM_THEN_REM)\n",
+                              pVCpu->em.s.enmState, enmOldState, EMSTATE_IEM_THEN_REM));
                         if (pVCpu->em.s.enmState != EMSTATE_IEM_THEN_REM)
                         {
                             pVCpu->em.s.enmState = EMSTATE_IEM_THEN_REM;
