charset.patch

From: Christian Prochaska <christian.prochaska@genode-labs.com>


---
 libcpp/charset.cc |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/libcpp/charset.cc b/libcpp/charset.cc
index ca8b7cf..b76d0fa 100644
--- a/libcpp/charset.cc
+++ b/libcpp/charset.cc
@@ -2480,9 +2480,10 @@ cpp_display_column_to_byte_column (const char *data, int data_length,
    are not highly optimized, but for the present purpose of outputting
    diagnostics, they are sufficient.  */
 
-#include "generated_cpp_wcwidth.h"
+//#include "generated_cpp_wcwidth.h"
 int cpp_wcwidth (cppchar_t c)
 {
+#if 0
   if (__builtin_expect (c <= wcwidth_range_ends[0], true))
     return wcwidth_widths[0];
 
@@ -2506,5 +2507,6 @@ int cpp_wcwidth (cppchar_t c)
 
   if (__builtin_expect (begin != end, true))
     return wcwidth_widths[begin];
+#endif
   return 1;
 }
