charset.patch

From: Christian Prochaska <christian.prochaska@genode-labs.com>


---
 libcpp/charset.cc |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/libcpp/charset.cc b/libcpp/charset.cc
index 54d7b9e..8353d8d 100644
--- a/libcpp/charset.cc
+++ b/libcpp/charset.cc
@@ -3406,11 +3406,12 @@ get_cppchar_property (cppchar_t c,
    are not highly optimized, but for the present purpose of outputting
    diagnostics, they are sufficient.  */
 
-#include "generated_cpp_wcwidth.h"
+//#include "generated_cpp_wcwidth.h"
 
 int
 cpp_wcwidth (cppchar_t c)
 {
+#if 0
   const size_t num_ranges
     = sizeof wcwidth_range_ends / sizeof (*wcwidth_range_ends);
   return get_cppchar_property<unsigned char > (c,
@@ -3418,6 +3419,9 @@ cpp_wcwidth (cppchar_t c)
 					       &wcwidth_widths[0],
 					       num_ranges,
 					       1);
+#else
+  return 1;
+#endif
 }
 
 #include "combining-chars.inc"
