i386.patch

From: Christian Prochaska <christian.prochaska@genode-labs.com>


---
 gcc/config/i386/i386.h |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/gcc/config/i386/i386.h b/gcc/config/i386/i386.h
index 363082b..dc3e19d 100644
--- a/gcc/config/i386/i386.h
+++ b/gcc/config/i386/i386.h
@@ -2351,6 +2351,7 @@ constexpr wide_int_bitmask PTA_ALDERLAKE = PTA_TREMONT | PTA_ADX | PTA_AVX
 constexpr wide_int_bitmask PTA_KNM = PTA_KNL | PTA_AVX5124VNNIW
   | PTA_AVX5124FMAPS | PTA_AVX512VPOPCNTDQ;
 
+#if 0
 #ifndef GENERATOR_FILE
 
 #include "insn-attr-common.h"
@@ -2373,6 +2374,7 @@ extern unsigned int const pta_size;
 extern unsigned int const num_arch_names;
 #endif
 
+#endif
 #endif
 
 extern enum processor_type ix86_tune;
@@ -2847,11 +2849,13 @@ extern void debug_dispatch_window (int);
 
 #define TARGET_SUPPORTS_WIDE_INT 1
 
+#if 0
 #if !defined(GENERATOR_FILE) && !defined(IN_LIBGCC2)
 extern enum attr_cpu ix86_schedule;
 
 #define NUM_X86_64_MS_CLOBBERED_REGS 12
 #endif
+#endif
 
 /* __builtin_eh_return can't handle stack realignment, so disable MMX/SSE
    in 32-bit libgcc functions that call it.  */
