diff --git a/src/share/mk/bsd.lib.mk b/src/share/mk/bsd.lib.mk
index ab658e7..85575bb 100644
--- a/src/share/mk/bsd.lib.mk
+++ b/src/share/mk/bsd.lib.mk
@@ -402,7 +402,7 @@ _LIB_g.a:=${_LIB}_g.a
 _LIB_pic.a:=${_LIB}_pic.a
 _LIB.ln:=llib-l${LIB}.ln
 
-.if ${MKPIC} != "no" && defined(SHLIB_FULLVERSION)
+.if ${MKPIC} != "no" && defined(SHLIB_FULLVERSION) && 0
 _LIB.so:=${_LIB}.so
 _LIB.so.major:=${_LIB}.so.${SHLIB_MAJOR}
 _LIB.so.full:=${_LIB}.so.${SHLIB_FULLVERSION}
