diff -r 6978f825431c drivers/usb/host/xhci-plat.c
--- a/drivers/usb/host/xhci-plat.c	Thu May 16 15:54:34 2013 +0200
+++ b/drivers/usb/host/xhci-plat.c	Thu May 16 16:41:46 2013 +0200
@@ -25,6 +25,16 @@
 	 * dev struct in order to setup MSI
 	 */
 	xhci->quirks |= XHCI_BROKEN_MSI;
+
+#ifdef DWC3_QUIRK
+	/* 
+	 * DWC3 controller
+	 * One will see 'dev_error: ERROR Transfer event TRB DMA ptr not part of
+	 * current TD' and last message was short (ep_ring->last_td_was_short, in
+	 * xhci-ring.c
+	 */
+	xhci->quirks |= XHCI_SPURIOUS_SUCCESS;
+#endif
 }
 
 /* called during probe() after chip reset completes */
