diff --git a/jitterentropy-base.c b/jitterentropy-base.c
index 28b1462..3ea4c90 100644
--- a/jitterentropy-base.c
+++ b/jitterentropy-base.c
@@ -553,7 +553,7 @@ ssize_t jent_read_entropy(struct rand_data *ec, char *data, size_t len)
 			tocopy = (DATA_SIZE_BITS / 8);
 		else
 			tocopy = len;
-		memcpy(p, &ec->data, tocopy);
+		jent_memcpy(p, &ec->data, tocopy);
 
 		len -= tocopy;
 		p += tocopy;
