+++ src/lib/libc/lib/libc/gen/_pthread_stubs.c
@@ -253,7 +253,7 @@
 STUB_FUNC2(pthread_attr_setschedparam, PJT_ATTR_SETSCHEDPARAM, int, void *, void *)
 STUB_FUNC2(pthread_attr_setschedpolicy, PJT_ATTR_SETSCHEDPOLICY, int, void *, int)
 STUB_FUNC2(pthread_attr_setscope, PJT_ATTR_SETSCOPE, int, void *, int)
-STUB_FUNC1(pthread_cancel, PJT_CANCEL, int, void *)
+/* STUB_FUNC1(pthread_cancel, PJT_CANCEL, int, void *) */
 STUB_FUNC1(pthread_cleanup_pop, PJT_CLEANUP_POP, int, int)
 STUB_FUNC2(pthread_cleanup_push, PJT_CLEANUP_PUSH, void, void *, void *)
 STUB_FUNC3(pthread_cond_timedwait, PJT_COND_TIMEDWAIT, int, void *, void *, void *)
