--- contrib/dosbox-svn-3837/src/gui/sdl_mapper.cpp.orig	2013-10-30 13:57:07.000000000 +0100
+++ contrib/dosbox-svn-3837/src/gui/sdl_mapper.cpp	2013-10-30 15:35:50.563742200 +0100
@@ -416,7 +416,7 @@
 		} 
 #if !defined (WIN32) && !defined (MACOSX) && !defined(OS2)
 		/* Linux adds 8 to all scancodes */
-		else key-=8;
+		//else key-=8;
 #endif
 #if defined (WIN32)
 		switch (key) {
@@ -2449,6 +2449,8 @@
 			}
 		}
 #endif
+		/* Genode's input keycodes match these of evdev */
+		evdev_input = true;
 		if (evdev_input) {
 			sdlkey_map[0x67]=SDLK_UP;
 			sdlkey_map[0x6c]=SDLK_DOWN;
